text/0034-ternary-hash.md

```diff
@@ -0,0 +1,472 @@
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input provided by the user is `absorb`ed, and an output will be `squeeze`d from the
+data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into` versions of
+these methods are for passing a buffer into which the calculated hashes are written. The internal state will not be
+cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729`:
+
+```rust
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step:
+
+||-1|0|1|
+|-|-|-|-|
+|**-1**|1|1|-1|
+|**0**|0|-1|1|
+|**1**|-1|0|0|
+
+Given two balanced trits `t` and `t'`, the following table can easily be accessed by shifting them to unbalanced trits:
+```rust
+const TRUTH_TABLE: [[i8; 3]; 3] = [[1, 0, -1], [1, -1, 0], [-1, 1, 0]];
+```
+
+The way `CurlP` is defined, it can not actually fail, because the input or outputs can be of arbitrary size; hence,
+the associated type `Error = Infallible`.
+
+`CurlP` has two common variants depending on the number of rounds of hashing to apply before a hash is squeezed.
+```rust
+#[derive(Copy, Clone)]
+pub enum CurlPRounds {
+    Rounds27 = 27,
+    Rounds81 = 81,
+}
+```
+
+Type definition:
+
+```rust
+struct CurlP {
+    /// The number of rounds of hashing to apply before a hash is squeezed.
+    rounds: CurlPRounds,
+
+    /// The internal state.
+    state: TritBuf,
+
+    /// Workspace for performing transformations.
+    work_state: TritBuf,
+}
+```
+
+In addition, the two wrapper types for the very common `CurlP` variants with `27` and `81` rounds:
+
+```rust
+struct CurlP27(CurlP);
+
+impl CurlP27 {
+    pub fn new() -> Self {
+        Self(CurlP::new(CurlPRounds::Rounds27))
+    }
+}
+
+struct CurlP81(CurlP);
+
+impl CurlP81 {
+    pub fn new() -> Self {
+        Self(CurlP::new(CurlPRounds::Rounds81))
+    }
+}
+```
+
+For convenience, they should both dereference to an actual `CurlP`:
+
+```rust
+impl Deref for CurlP27 {
+    type Target = CurlP;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for CurlP27 {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Deref for CurlP81 {
+    type Target = CurlP;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for CurlP81 {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+```
+
+This allows using them as a `Sponge` as well if there is a blanket implementation like this:
+
+```rust
+impl<T: Sponge, U: DerefMut<Target = T>> Sponge for U {
```

Comment:
I mentioned it in a comment before: this blanket impl would be more logical, if the reader had seen a `impl Sponge for CurlP` before.